internal/federationdomain/oidc/oidc.go

```diff
@@ -104,26 +105,152 @@ type UpstreamStateParamData struct {
+	// that the storage be garbage collected in the middle of trying to look up the token.
+	storageExtraLifetime := time.Minute
+
+	// This is longer than the 5-minute lifetime of mTLS client certs issued by the Concierge,
+	// so this should allow a user to fetch another client cert after the first one expires,
+	// without needing to refresh their Supervisor session.
+	gitHubPATBasedAccessTokenLifespan := 8 * time.Minute
+
+	// Previous versions of the Pinniped CLI would only skip refresh when there is at least
+	// 10 minutes left for the cached ID token, so having a short lifetime here will cause those
+	// older CLIs to never skip attempting refresh. Refreshes are not allowed for GitHub PAT-based
+	// sessions, so those older CLIs will always get a refresh failure and then automatically start
+	// a new session. This is unfortunate because it uses more of a user's GitHub API rate limit
+	// per hour, and it uses more Supervisor session storage (more new sessions). However, we don't
+	// want to make this lifetime long because this is also the lifetime of cluster-scoped ID tokens,
+	// which can grant access to clusters, so we will have to live with this. Users can avoid it by
+	// upgrading their CLI because newer CLIs will look at the lifespan of the access token (not the
+	// ID token) when deciding if a refresh is needed before requesting a new cluster-scoped ID token.
+	gitHubPATBasedIDTokenLifespan := idTokenLifespan
+
+	// Give a little extra time for some storage lifetimes, to avoid the possibility
+	// that the storage be garbage collected in the middle of trying to look up the token.
+	gitHubPATBasedTokenStorageExtraLife := 5 * time.Second
+
 	return timeouts.Configuration{
-		UpstreamStateParamLifespan:              90 * time.Minute,
-		AuthorizeCodeLifespan:                   authorizationCodeLifespan,
-		AccessTokenLifespan:                     accessTokenLifespan,
-		IDTokenLifespan:                         accessTokenLifespan,
-		RefreshTokenLifespan:                    refreshTokenLifespan,
-		AuthorizationCodeSessionStorageLifetime: authorizationCodeLifespan + refreshTokenLifespan,
-		PKCESessionStorageLifetime:              authorizationCodeLifespan + (1 * time.Minute),
-		OIDCSessionStorageLifetime:              authorizationCodeLifespan + (1 * time.Minute),
-		AccessTokenSessionStorageLifetime:       refreshTokenLifespan + accessTokenLifespan,
-		RefreshTokenSessionStorageLifetime:      refreshTokenLifespan + accessTokenLifespan,
+		// Give enough time for someone to start an interactive authorization flow, go eat lunch,
+		// and then finish the authorization afterward.
+		UpstreamStateParamLifespan: 90 * time.Minute,
+
+		AuthorizeCodeLifespan: authorizationCodeLifespan,
+
+		AccessTokenLifespan: accessTokenLifespan,
+		OverrideDefaultAccessTokenLifespan: func(accessRequest fosite.AccessRequester) (bool, time.Duration) {
+			if isGitHubSessionBasedOnPAT(accessRequest) {
```

Comment:
The "if GithubXYZ..." is pretty clean.  Working off my above comment, I imagine this is easy enough to maintain until we have several that require deviance.

Reply:
👍